src/aead/chacha20_poly1305.rs

```diff
@@ -256,29 +244,111 @@ extern {
 
 #[cfg(test)]
 mod tests {
-    use {aead, c, polyfill, test};
-    use super::{ChaCha20_ctr32, CHACHA20_KEY_LEN, make_counter};
-
-    bssl_test!(test_poly1305, bssl_poly1305_test_main);
+    use {aead, test, error, c, polyfill};
+    use super::{ChaCha20_ctr32, CHACHA20_KEY_LEN, POLY1305_STATE_LEN, POLY1305_KEY_LEN,
+                poly1305_init, poly1305_update, poly1305_finish, make_counter};
 
     #[test]
     pub fn test_chacha20_poly1305() {
         aead::tests::test_aead(&aead::CHACHA20_POLY1305,
-            "crypto/cipher/test/chacha20_poly1305_tests.txt");
+                               "crypto/cipher/test/chacha20_poly1305_tests.txt");
     }
 
     #[test]
     pub fn test_chacha20_poly1305_old() {
         aead::tests::test_aead(&aead::CHACHA20_POLY1305_OLD,
-            "crypto/cipher/test/chacha20_poly1305_old_tests.txt");
+                               "crypto/cipher/test/chacha20_poly1305_old_tests.txt");
     }
 
     #[test]
     pub fn test_poly1305_state_len() {
-        assert_eq!((super::POLY1305_STATE_LEN + 255) / 256,
-                    (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
+        assert_eq!((POLY1305_STATE_LEN + 255) / 256,
+                   (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
+    }
+
+    fn test_simd(excess: usize, key: [u8; POLY1305_KEY_LEN],
+                 input: &[u8], mac: [u8; aead::TAG_LEN])
+                 -> Result<(), error::Unspecified> {
+        let mut state = [0u8; POLY1305_STATE_LEN];
+        let mut mac_out = [0u8; aead::TAG_LEN];
+        poly1305_init(&mut state, &key);
+
+        // Feed 16 bytes in. Some implementations begin in non-SIMD mode and
+        // upgrade on-demand. Stress the upgrade path.
+        let init = if input.len() < 16 { input.len() } else { 16 };
+
+        poly1305_update(&mut state, &input[..init]);
+        for chunk in input[init..].chunks(128 + 2 * excess) {
+            let (long, short) = if chunk.len() < (128 + excess) {
+                (chunk, &[][..])
+            } else {
+                chunk.split_at(128 + excess)
+            };
+            // Feed 128 + |excess| bytes to test SIMD mode
+            poly1305_update(&mut state, long);
+            // Feed |excess| bytes to ensure SIMD mode can handle short inputs
+            if !short.is_empty() {
+                poly1305_update(&mut state, short);
+            }
+        }
+        poly1305_finish(&mut state, &mut mac_out);
```

Comment:
Let's move the declaration/initialization of `mac_out` here. I suggest renaming `mac` to `expected_mac` and rename `mac_out` to `actual_mac`.